### PR TITLE
Feature/refactor api

### PR DESCRIPTION
**Pull request by Nainterceptor**

### CODE REVIEW

File: api/abstract/campus-crud-query.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  get campusRoutePrefix() {
+    if (!this.campus) {
+      return null;
+    }
+    return `/campuses/${encodeURIComponent(this.campus)}`;
+  }
+
+  getEndpoint(...params) {
+    const base = [this.constructor.baseEndpoint, ...params.map((f) => encodeURIComponent(f))].join('/');
+    if (!this.campus) {
```
Comment (Scttpr) [inline api/abstract/campus-crud-query.js:10]: Could end up as a ternary

File: api/abstract/query.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+  setMask(mask) {
+    if (Array.isArray(mask)) {
+      this.mask = mask.join(',');
+      return this;
+    }
+    this.mask = mask;
+    return this;
+  }
```
Comment (Scttpr) [inline api/abstract/query.js:22]:
```suggestion
  setMask(mask) {
  this.mask = Array.isArray(mask) ? this.mask = mask.join(',') : mask;
  }
```

File: api/car-models.js [hunk 3]
```diff
@@ -1,80 +1,14 @@
...
-      params.csv = csv;
-    }
-    const response = await axios.get(
-      `/${ENTITY_PLURAL}`,
-      {
-        params,
-        headers,
-      },
-    );
+export default class CampusesQuery extends AbstractCRUDQuery {
```
Comment (Scttpr) [inline api/car-models.js:6]:
```suggestion
export default class CarModelsQuery extends AbstractCRUDQuery {
```

File: api/cars.js [hunk 4]
```diff
@@ -1,88 +1,14 @@
...
+import AbstractCampusCRUDQuery from './abstract/campus-crud-query';
 
 export const ENTITY = 'car';
 export const ENTITY_PLURAL = 'cars';
 
-export default (axios) => (campus, mask) => {
-  const filters = {};
-  if (campus) {
-    filters.campus = campus.id;
+export default class PoisQuery extends AbstractCampusCRUDQuery {
```
Comment (Scttpr) [inline api/cars.js:6]:
```suggestion
export default class CarsQuery extends AbstractCampusCRUDQuery {
```

File: api/drivers.js [hunk 5]
```diff
@@ -1,91 +1,14 @@
...
 
 export const ENTITY_PLURAL = 'drivers';
-export const ENTITY_CAMPUSES = 'campuses';
+export const ENTITY = 'user';
 
-export default (axios) => (campus, mask) => {
-  const filters = {};
-  if (campus) {
-    filters.campus = campus.id;
+export default class PoisQuery extends AbstractCampusCRUDQuery {
```
Comment (Scttpr) [inline api/drivers.js:6]:
```suggestion
export default class DriversQuery extends AbstractCampusCRUDQuery {
```

File: api/rides.js [hunk 6]
```diff
@@ -1,237 +1,147 @@
...
+    return this.constructor.axios.get(
+      `${this.campusRoutePrefix || ''}/stats`,
+      {
+        params: {
+          mask: this.mask,
+          filters: {
+            start,
+            end,
+            'time-scope': options.timeScope,
+            'time-unit': options.timeUnit,
```
Comment (Scttpr) [inline api/rides.js:107]:
```suggestion
  async stats(start, end, options = {}) {
    const filters = { start, end };
    if (options['time-scope']) {
         filters['time-scope'] = options['time-scope'];
    }
        if (options['time-unit']) {
         filters['time-unit'] = options['time-unit'];
    }
    return this.constructor.axios.get(
      `${this.campusRoutePrefix || ''}/stats`,
      {
        params: {
          mask: this.mask,
          filters,
```

File: api/time-slots.js [hunk 7]
```diff
@@ -1,74 +1,29 @@
...
-        merge(data, { campus: { id: campus } }),
-        {
-          params,
-        },
-      );
-    },
+  list(after, before, options = {}) {
+    return super.list(options)
+      .setFilter('after', after)
+      .setFilter('before', before);
```
Comment (Scttpr) [inline api/time-slots.js:17]: I Would set `before` before `after`

File: api/users.js [hunk 8]
```diff
@@ -1,98 +1,46 @@
...
-  patchUser(id, data, mask, { sendToken = false } = {}) {
-    return axios.patch(
-      `/${ENTITY_PLURAL}/${encodeURIComponent(id)}`,
-      data,
-      {
-        params: { mask },
-        headers: {
-          'X-Send-Token': sendToken,
-        },
+  create(data, options) {
```
Comment (Scttpr) [inline api/users.js:21]: No default value here ? It is set in others similar cases, should not throw errors though.

File: api/users.js [hunk 9]
```diff
@@ -1,98 +1,46 @@
...
-  postUser(data, mask, { sendToken = false } = {}) {
-    return axios.post(
-      `/${ENTITY_PLURAL}`,
-      data,
-      {
-        params: { mask },
-        headers: {
-          'X-Send-Token': sendToken,
-        },
+  edit(id, data, options) {
```
Comment (Scttpr) [inline api/users.js:29]: No default value here ? It is set in others similar cases, should not throw errors though.

File: api/users.js [hunk 10]
```diff
@@ -1,98 +1,46 @@
...
       },
-    );
-  },
+    }, options)));
+  }
 
-  postUsers(data, { sendToken = false, delimiter = ';', ignoreEmpty = true } = {}) {
-    return axios.post(
-      `/${ENTITY_PLURAL}/batch`,
+  subscribeDevice(id, data, options) {
```
Comment (Scttpr) [inline api/users.js:37]: No default value here ? It is set in others similar cases, should not throw errors though.

File: api/rides.js (hunk 6 above)
Comment (Nainterceptor) [reply api/rides.js:107]: Options should to be merged with axios options, on JSON.stringify, undefined properties are dropped

**Pull request merged**